File: src/main/scala/imcode/server/document/index/service/impl/IndexRebuildScheduler.scala
package imcode.server.document.index.service.impl

import com.imcode._
import com.imcode.util.Threads
import java.util.concurrent._
import java.util.concurrent.atomic.AtomicReference
import imcode.server.document.index.service.DocumentIndexService


trait IndexRebuildScheduler { this: DocumentIndexService =>

  private val executor = new ScheduledThreadPoolExecutor(1, new ThreadFactory {
    def newThread(runnable: Runnable): Thread = new Thread(runnable) |>> { thread =>
      thread.setName("index-rebuild-scheduler")
      thread.setDaemon(true)
    }
  })

  private val scheduledFutureRef = new AtomicReference[ScheduledFuture[_]]

  /**
   * Schedules periodic index rebuild with fixed interval.
   * Cancels any previously scheduled rebuild.
   * If interval is not given or its value is less or equals to zero then rebuild is not scheduled.
   *
   * @param intervalOpt interval in minutes.
   */
  def setRebuildIntervalInMinutes(intervalOpt: Option[Int]): Unit = synchronized {
    scheduledFutureRef.getAndSet(null).asOption.foreach(_.cancel(true))

    for (interval <- intervalOpt if interval > 0) {
      logger.info(s"Index rebuild schedule interval is set to $interval minutes.")

      executor.scheduleWithFixedDelay(
        Threads.mkRunnable {
          logger.info("Running scheduled index rebuild.")
          rebuild().foreach { indexRebuildTask =>
            try {
              indexRebuildTask.future.get
              logger.info(s"Scheduled index rebuild task has finished. Next scheduled run in $interval minutes.")
            } catch {
              case _: InterruptedException =>
                logger.info("Scheduled index rebuild task has been interrupted.")

              case _: CancellationException =>
                logger.info("Scheduled index rebuild task has been cancelled.")

              case e: ExecutionException =>
                logger.info("Scheduled index rebuild task has failed.", e)
            }
          }
        },
        interval,
        interval,
        TimeUnit.MINUTES
      ) |> scheduledFutureRef.set
    }
  }
}

File: src/main/java/imcode/server/document/index/service/impl/IndexRebuildScheduler.java
package imcode.server.document.index.service.impl;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicReference;

// translated from scala...
public interface IndexRebuildScheduler {

    Logger logger = LogManager.getLogger(IndexRebuildScheduler.class);

    ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1, runnable -> {
        final Thread thread = new Thread(runnable);
        thread.setName("index-rebuild-scheduler");
        thread.setDaemon(true);

        return null;
    });

    AtomicReference<ScheduledFuture> scheduledFutureRef = new AtomicReference<>();

    /**
     * Schedules periodic index rebuild with fixed interval.
     * Cancels any previously scheduled rebuild.
     * If interval is not given or its value is less or equals to zero then rebuild is not scheduled.
     *
     * @param interval interval in minutes.
     */
    default void setRebuildIntervalInMinutes(long interval) {
        final ScheduledFuture future = scheduledFutureRef.getAndSet(null);

        if (future != null) {
            future.cancel(true);
        }

        if (interval > 0) {
            logger.info("Index rebuild schedule interval is set to $interval minutes.");

            final Runnable scheduledCommand = () -> {
                logger.info("Running scheduled index rebuild.");

                try {
                    rebuild().get();
                    logger.info("Scheduled index rebuild task has finished. Next scheduled run in " + interval + " minutes.");

                } catch (InterruptedException e) {
                    logger.info("Scheduled index rebuild task has been interrupted.");
                } catch (CancellationException e) {
                    logger.info("Scheduled index rebuild task has been cancelled.");
                } catch (ExecutionException e) {
                    logger.info("Scheduled index rebuild task has failed.", e);
                }
            };

            final ScheduledFuture<?> scheduledFuture = executor.scheduleWithFixedDelay(
                    scheduledCommand, interval, interval, TimeUnit.MINUTES
            );

            scheduledFutureRef.set(scheduledFuture);
        }
    }

    Future rebuild();
}
